### DISCUSSION TITLE
Is nufft operation differentiable using torchkbnufft?

### DISCUSSION THREAD

**Discussion by Michaelsqj** [mmuckley/torchkbnufft#63] [Q&A]
Dear Matthew,

I'm having a 3D non-Cartesian (cone) trajectory, and I want to optimize for the rotation angle for each individual spoke. After initializing a set of rotation angles, calculate the density compensation weighting $w$ using `calc_density_compensation_function `, and then calculate a metrics $\frac{\sum w}{\sum\sqrt{w^2}}$ to evaluate the uniformity of the weighting as the cost function. The rotation angles are optimized by maximizing the uniformity. Do you think it's possible to do such optimization using torchkbnufft?

Yours,
Qijia

**Comment by mmuckley** [comment]
Hello @Michaelsqj, as mentioned in #62, `torchkbnufft` does not compute gradients with respect to the trajectory. There is a rounding operation that we use to construct the interpolation table, and this can't be differentiated through. Doing direct trajectory optimization has been the subject of some recent papers, such as in the following work:

Code: https://github.com/guanhuaw/Bjork
Paper: https://arxiv.org/abs/2101.11369

I might recommend that you look there.

**Reply by Michaelsqj** [reply]
Thank you! Let me check the repos you pointed to.

**Reply by Michaelsqj** [reply]
Hello @mmuckley , I took closer look at your code and found `class KbTableInterpAdjoint(Function)` implemented both forward and backward method https://github.com/mmuckley/torchkbnufft/blob/3e081be6cc0aa866829070882f3f7e5a7d111c93/torchkbnufft/_autograd/interp.py#L133, and the comment also said `This is a wrapper for for PyTorch autograd.` So what is the purpose of these method if `kb_table_interp_adjoint` function is not differentiable?

Or is the `kb_table_interp_adjoint` can be differentiated but nufft couldn't?

**Reply by mmuckley** [reply]
@Michaelsqj you can backpropagate to image-domain weights just fine. You cannot backpropagate to k-space trajectory locations. This is due to the underlying algorithms.
